scripts/mods/armory/armory.lua

```diff
@@ -104,6 +104,40 @@ mod:hook_safe(StateInGameRunning, "update", function (self)
     end
 end)
 
+local function get_upvalue_by_name(obj, upvalue_name)
+    for _, v in pairs(obj) do
+        if type(v) == "function" then
+            local debug_info = debug.getinfo(v)
+            local func = debug_info.func
+            local upvalues_n = debug_info.nups
+
+            for i = 1, upvalues_n do
+                local up_name, up_value = debug.getupvalue(func, i)
+                if up_name == upvalue_name then
+                    mod:echo(up_value)
+                    return up_value, i, v
+                end
+            end
+
+        end
+    end
+end
+
+local function set_upvalue_by_name(obj, upvalue_name, new_value)
+    local upvalue, upvalue_index, func = get_upvalue_by_name(obj, upvalue_name)
+    debug.setupvalue(func, upvalue_index, new_value)
+end
+
+mod:hook(Unit, "get_data", function(func, unit, ...)
+    if unit == nil or not unit then
+        return
+    end
+
+    return func(unit, ...)
+end)
+
+set_upvalue_by_name(ActionUtils, "unit_get_data", Unit.get_data)
```

Comment:
Unit.get_data is used a lot throughout the code base. Hooking into it will lead to a lot of memory allocations and in turn performance issues. I would much rather prefer we solve the cases where unit is passed in as nil.

This may also mask potential crashes related to get_data, causing them to be difficult to track.

Reply:
I fully agree with preferring a different route. I went this way because `ActionUtils.get_power_level_for_target` added a check for an armor override on enemy units; however, I'm using `DamageUtils.calculate_damage_tooltip`, which uses the same `do_damage_calculation` function but doesn't supply a `target_unit`.

My ideal would be Fatshark adding a nil check for `target_unit` before checking for an override, because I feel the current implementation also breaks the `DamageUtils.calculate_damage_tooltip` method (which I realize may be an artifact from V1).

I could get around the hook by instead replacing the `unit_get_data` upvalue with a simple function that checks for nil and returns early, but this wouldn't address your second concern.

Reply:
Or I could do something like this:
```lua
local unit_get_data = get_upvalue_by_name(ActionUtils, "unit_get_data")
set_upvalue_by_name(ActionUtils, "unit_get_data", function(unit, ...)
    if not unit and mod.mock_damage then
        return
    end

    return unit_get_data(unit, ...)
end)
```

and toggle `mod.mock_damage` before/after mod-related damage checks.

Reply:
Ah, that's right. The breaking changes are a miss on our end. If the delay is fine with you I have a fix lined up for a hotfix next week. I can keep you posted on it.

Otherwise I'm fine with your last proposed solution if you'd rather get this done asap.

Reply:
If you've a fix slated for a hotfix, I think that's preferable. While not as much overhead as a hook, a wrapper still incurs a cost.

Reply:
Fixes are live.